File: implementations/rust/src/espprc.rs
use std::{cell::Cell};

use crate::tsp;
use std::collections::VecDeque;

struct Label {
    at: usize,
    visited: Vec<bool>,
    ignore: Cell<bool>,
    _predecessor: Option<usize>,
    cost: f64,
    length: i32,
    q: Vec<usize>,
    successors: Vec<usize>,
}

impl Label {
    fn dominates(&self, other: &Label) -> bool {
        if self.cost > other.cost || self.length > other.length {
            return false;
        }
        for (v1, v2) in self.visited.iter().zip(other.visited.iter()) {
            if v1 > v2 { return false}
        }
        for (q1, q2) in self.q.iter().zip(other.q.iter()) {
            if q1 > q2 { return false}
        }
        return true;
    }
}

struct LabelCollection {
    labels: Vec<Label>,
}

impl LabelCollection {
    fn empty_label(&mut self, d: &tsp::TSPData, nresources: usize) -> usize {
        let visited: Vec<bool> = vec![false; d.n];
        let q: Vec<usize> = vec![0; nresources];
        self.labels.push(Label {
            at: 0,
            visited: visited,
            ignore: Cell::new(false),
            _predecessor: None,
            cost: 0.0,
            length: 0,
            q: q,
            successors: Vec::new(),
        });
        return self.labels.len() - 1;
    }

    // from is the index of a Label
    // to is a node index
    fn extend(&mut self, d: &tsp::TSPData, from: usize, to: usize) -> usize {
        let mut visited = self.labels[from].visited.clone();
        visited[to] = true;
        let mut q = self.labels[from].q.clone();
        for i in 0..q.len() {
            if (to & (1 << i)) > 0 {
                q[i] += 1;
            }
        }
        let cost = self.labels[from].cost + d.aux(self.labels[from].at, to);
        let length = self.labels[from].length + d.d(self.labels[from].at, to);
        let nl = Label{at: to,
                       visited: visited,
                       ignore: Cell::new(false), 
                       _predecessor: Some(from),
                       cost: cost,
                       length: length,
                       q: q,
                       successors: Vec::new()};
        self.labels.push(nl);
        return self.labels.len() - 1;
    }

    // add label to as successor to label from
    fn addsuccessor(&mut self, from: usize, to: usize) {
        self.labels[from].successors.push(to);
    }
    
    fn marksuccessors(&self, index: usize) {
        let Label{ successors, .. } = &self.labels[index];
        for &s in successors {
            &self.labels[s].ignore.set(true);
            self.marksuccessors(s);
        }
    }

    fn updatedominance(&mut self, labels: &mut Vec<usize>, nlindex: usize) -> bool {
        let mut i: usize = 0;
        while i < labels.len() {
            if self.labels[labels[i]].dominates(&self.labels[nlindex]) {
                return false;
            }
            if self.labels[nlindex].dominates(&self.labels[labels[i]]) {
                self.marksuccessors(labels[i]);
                if i < labels.len() - 1 {
                    labels[i] = *labels.last().unwrap();
                }
                labels.pop();
            } else {
                i += 1;
            }
        }
        // at this point the new label is not dominated so we add it
        labels.push(nlindex);
        return true;
    }
}

pub fn solve(d: &tsp::TSPData,
             nresources: usize, resourcecapacity: usize, maxlen: i32) -> i32 {
    // we will store all labels here
    let mut lc = LabelCollection{ labels: Vec::new() };
    let mut q: VecDeque<usize> = VecDeque::new();
    let mut in_q: Vec<bool> = vec![false; d.n];
    // initial label
    let l0 = lc.empty_label(d, nresources);
    q.push_back(0);
    in_q[0] = true;
    // considered labels at each node
    let mut labels: Vec<Vec<usize>> = vec![Vec::new(); d.n];
    labels[0].push(l0);
    // main DP loop
    while ! q.is_empty() {
        let n = q.pop_front().unwrap();
        in_q[n] = false;
        for i in 0..labels[n].len() {
            let lind = labels[n][i];
            if lc.labels[lind].ignore.get() == true {
                continue;
            }
            for succ in 0..d.n {
                if lc.labels[lind].visited[succ] || succ == n { continue }
                // is the extension length-feasible?
                if lc.labels[lind].length + d.d(n, succ) + d.d(succ, 0)
                    > maxlen {
                        continue;
                    }
                // is it resource-feasible?
                let mut rfeas: bool = true;
                for r in 0..nresources {
                    if (succ & (1 << r)) > 0 &&
                        lc.labels[lind].q[r] + 1 > resourcecapacity {
                            rfeas = false;
                            break;
                        }
                }
                if ! rfeas { continue }
                // at this point we know the extension is feasible
                let nl = lc.extend(d, lind, succ);
                let added = lc.updatedominance(&mut labels[succ], nl);
                if added {
                    lc.addsuccessor(lind, nl);
                    if ! in_q[succ] && succ != 0 {
                        q.push_back(succ);
                        in_q[succ] = true;
                    }
                }
            }
            lc.labels[lind].ignore.set(true);
        }
    }
    let mut bestcost: f64 = lc.labels[labels[0][0]].cost;
    for i in 1..labels[0].len() {
        if lc.labels[labels[0][i]].cost < bestcost {
            bestcost = lc.labels[labels[0][i]].cost;
        }
    }
    return bestcost as i32;
}

File: implementations/java/ESPPRC.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.ArrayDeque;

class Label {
    // where we're at
    private int at_;
    // already visited nodes
    private boolean[] visited_;
    // has this label already been extended?
    private boolean ignore_;
    // predecessor label
    private Label pred_;
    //
    private double cost_;
    private int length_;
    // resource consumption
    private int[] q_;
    // successors for later recursive deletion when dominated
    private ArrayList<Label> successors_;


    // number of resources considered
    public static int nresources;

    // number of vertices (also gives dimension of cost/len matrix)
    public static int nnodes;
    
    public Label() {
        at_ = 0;
        visited_ = new boolean[nnodes];
        Arrays.fill(visited_, false);
        ignore_ = false;
        pred_ = null;
        cost_ = 0.0;
        length_ = 0;
	q_ = new int[nresources];
        Arrays.fill(q_, 0);
        successors_ = new ArrayList<Label>();
    }
    
    private boolean dominates (Label other) {
	if (cost_ > other.cost_ || length_ > other.length_) return false;
	for (int i=0; i < nnodes; i++) {
	    if (visited_[i] && ! other.visited_[i]) return false;
	}
	for (int i=0; i < nresources; i++) {
	    if (q_[i] > other.q_[i]) return false;
	}
	return true;
    }

    // constructor by extension
    public Label(Label pred, int vertex, double[] rc, int[] d) {
	ignore_ = false;
	at_ = vertex;
	pred_ = pred;
	visited_ = new boolean[nnodes];
        System.arraycopy(pred.visited_, 0, visited_, 0, nnodes);
	visited_[vertex] = true;
	cost_ = pred.cost_ + rc[pred.at_ * nnodes + vertex];
	length_ = pred.length_ + d[pred.at_ * nnodes + vertex];
	q_ = new int[nresources];
        System.arraycopy(pred_.q_, 0, q_, 0, nresources);
	for (int r=0; r < nresources; r++) {
	    if ((vertex & (1 << r)) > 0) {
		q_[r] += 1;
	    }
	}
        successors_ = new ArrayList<Label>();
    }

    protected void addsuccessor(Label successor) {
	successors_.add(successor);
    }

    // getters
    protected boolean ignore() { return ignore_; }
    protected void setignore() { ignore_ = true; }
    protected boolean visits(int n) { return visited_[n]; }
    protected int length()  { return length_; }
    protected double cost() { return cost_; }
    private int at() { return at_; }
    protected int getrusage(int r) { return q_[r]; }
    Label predecessor() { return pred_; }
    ArrayList<Label> successors() { return successors_; }
    
    void marksuccessors() {
	for (int s=0; s < successors_.size(); s++) {
	    successors_.get(s).ignore_ = true;
	    successors_.get(s).marksuccessors();
	}
    }

    // update labels with newlabel, i.e. remove dominated elements as needed.
    // return true if newlabel is added, false otherwise
    protected static boolean update(ArrayList<Label> labels, Label newlabel) {
	int i=0;
	while (i < labels.size()) {
	    if (labels.get(i).dominates(newlabel)) {
		return false;
	    }
	    if (newlabel.dominates(labels.get(i))) {
		labels.get(i).marksuccessors();
		if (i < labels.size() - 1) {
		    labels.set(i, labels.get(labels.size()-1));
		}
		labels.remove(labels.size()-1);
	    } else {
		i += 1;
	    }
	}
	// at this point newlabel is not dominated so we add it
	labels.add(newlabel);
	return true;
    }
}

public class ESPPRC {

    protected int nnodes_;
    // reduced cost flat matrix
    protected double[] rc_;
    // length flat matrix
    protected int[] d_;
    // number of resources considered for resource constraints
    protected int nresources_;
    // capacity for each resource
    protected int resourcecapacity_;
    // maximum length for any given tour
    protected int maxlen_;
    
    public ESPPRC(int nnodes, double[] rc, int[] d,
                  int nresources, int rescap, int maxlen) {
	nnodes_ = nnodes;
	d_ = d;
	rc_ = rc;
	nresources_ = nresources;
	resourcecapacity_ = rescap;
	maxlen_ = maxlen;
    }

    public int solve(){
	// set class variables
	Label.nresources = nresources_;
	Label.nnodes = nnodes_;
	// initial label (empty path) and queue
	Label l = new Label();
	ArrayDeque<Integer> Q = new ArrayDeque<Integer> ();
	boolean[] inQ = new boolean[nnodes_];
        Arrays.fill(inQ, false);
	Q.addLast(0);
	inQ[0] = true;
        ArrayList<ArrayList<Label>> labels =
            new ArrayList<ArrayList<Label>>(nnodes_);
        for (int i=0; i < nnodes_; i++) {
            labels.add(new ArrayList<Label>());
        }
	labels.get(0).add(l);
	// main DP loop
	while (Q.size() > 0) {
	    int n = Q.pollFirst();
	    inQ[n] = false;
	    for (Label label: labels.get(n)) {
		if (label.ignore()) {
		    continue;
		}
		for (int succ = 0; succ < nnodes_; succ++) {
		    if (label.visits(succ) || succ == n) {
			continue;
		    }
                    // succ is a candidate for extension; is it length-feasible?
                    if (label.length()
			+ d_[n*nnodes_+succ] + d_[succ*nnodes_] > maxlen_) {
			continue;
		    }
		    // is it resource-feasible ?
		    boolean rfeas = true;
		    for (int i=0; i < nresources_; i++) {
			if ( (succ & (1 << i)) > 0 &&
			     label.getrusage(i) + 1 > resourcecapacity_ ) {
			    rfeas = false;
			    break;
			}
		    }
		    if (! rfeas) {
			continue;
		    }
		    // at this point we know the extension is feasible
		    Label nl = new Label(label, succ, rc_, d_);
		    boolean added = Label.update(labels.get(succ), nl);
		    if (added) label.addsuccessor(nl);
		    if (added && (! inQ[succ]) && succ != 0) {
			Q.addLast(succ);
			inQ[succ] = true;
		    }
		}
		label.setignore();
	    }
	}
	double bestcost = labels.get(0).get(0).cost();
	for (int i=1; i < labels.get(0).size(); i++) {
	    if (labels.get(0).get(i).cost() < bestcost) {
		bestcost = labels.get(0).get(i).cost();
	    }
	}
	return (int) bestcost;
    }
}
